ydb/core/persqueue/partition.cpp

```diff
@@ -1649,127 +1655,206 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
         ScheduleDeletePartitionDone();
 
-        AddCmdDeleteRangeForAllKeys(*request);
-
-        ctx.Send(Tablet, request.Release());
+        AddCmdDeleteRangeForAllKeys(*PersistRequest);
 
+        ctx.Send(Tablet, PersistRequest.Release());
+        PersistRequest = nullptr;
         DeletePartitionState = DELETION_IN_PROCESS;
         KVWriteInProgress = true;
 
         return;
     }
-
-    HaveWriteMsg = false;
-
-    if (UserActionAndTransactionEvents.empty()) {
-        const auto now = ctx.Now();
-
-        if (ManageWriteTimestampEstimate) {
-            WriteTimestampEstimate = now;
+    while (true) {
+        if (BatchingState == ETxBatchingState::PreProcessing) {
+            ContinueProcessTxsAndUserActs(ctx);
         }
-
-        bool haveChanges = CleanUp(request.Get(), ctx);
-        if (DiskIsFull) {
-            AddCheckDiskRequest(request.Get(), NumChannels);
-            haveChanges = true;
+        if (BatchingState == ETxBatchingState::PreProcessing) {
+            return; // Still preprocessing - waiting for something;
         }
 
-        ProcessReserveRequests(ctx);
-
-        if (haveChanges || TxIdHasChanged || !AffectedUsers.empty() || ChangeConfig) {
-            AddCmdWriteTxMeta(request->Record);
-            AddCmdWriteUserInfos(request->Record);
-            AddCmdWriteConfig(request->Record);
+        // Preprocessing complete;
+        if (CurrentBatchSize > 0) {
+            Send(SelfId(), new TEvPQ::TEvTxBatchComplete(CurrentBatchSize));
+        }
+        CurrentBatchSize = 0;
 
-            WriteCycleStartTime = now;
-            WriteStartTime = now;
-            TopicQuotaWaitTimeForCurrentBlob = TDuration::Zero();
-            PartitionQuotaWaitTimeForCurrentBlob = TDuration::Zero();
-            WritesTotal.Inc();
-            AddMetaKey(request.Get());
-            ctx.Send(Tablet, request.Release());
-            KVWriteInProgress = true;
-            HaveWriteMsg = true;
+        if (UserActionAndTxPendingCommit.empty()) {
+            // Processing stopped and nothing to commit - finalize
+            BatchingState = ETxBatchingState::Finishing;
         } else {
-            AnswerCurrentWrites(ctx);
-            AnswerCurrentReplies(ctx);
+            // Process commit queue
+            ProcessCommitQueue();
         }
+        if (!UserActionAndTxPendingCommit.empty()) {
+            // Still pending for come commits
+            return;
+        }
+        // Commit queue processing complete. Now can either swith to persist or continue preprocessing;
+        if (BatchingState == ETxBatchingState::Finishing) { // Persist required;
+            RunPersist();
+            return;
+        }
+        BatchingState = ETxBatchingState::PreProcessing;
+    }
+}
 
+void TPartition::ContinueProcessTxsAndUserActs(const TActorContext&)
+{
+    Y_ABORT_UNLESS(!KVWriteInProgress);
+
+    if (WriteCycleSizeEstimate >= MAX_WRITE_CYCLE_SIZE || WriteKeysSizeEstimate >= MAX_USERS) {
```

Comment:
Rename MAX_USERS to MAX_KEYS. and change it to 10k. And check that this limit is bigger than limit for data blobs in supportive partition. There should be some limit in it.